Can `reset_index` return None?

<!-- Please include a self-contained copy-pastable example that generates the issue if possible.

Please be concise with code posted. See guidelines below on how to provide a good bug report:

- Craft Minimal Bug Reports: http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports
- Minimal Complete Verifiable Examples: https://stackoverflow.com/help/mcve

Bug reports that follow these guidelines are easier to diagnose, and so are often handled much more quickly.
-->

**What happened**:

The line
```
img = img.reset_index(["spatial"])
```
was not accepted by mypy. Mypy gave the following error:
```
redacted.py:69: error: Incompatible types in assignment (expression has type "Optional[DataArray]", variable has type "DataArray")
```
**What you expected to happen**:

I think that the return type of `DataArray.reset_index` should be `DataArray`, not `Optional[DataArray]`. `Optional[Array]` also includes `None` but the source code doesn't return None anywhere as far as I can see. Specifically, `reset_index` just redirects to `_replace` which returns a `DataArray` not `Optional[DataArray]` according to its type annotation.

**Relevant code in `xarray.core.dataarray.py`**:

```python
class DataArray(AbstractArray, DataWithCoords, DataArrayArithmetic):
.
.
.
    def _replace(
        self,
        variable: Variable = None,
        coords=None,
        name: Union[Hashable, None, Default] = _default,
        indexes=None,
    ) -> "DataArray":
        if variable is None:
            variable = self.variable
        if coords is None:
            coords = self._coords
        if name is _default:
            name = self.name
        return type(self)(variable, coords, name=name, fastpath=True, indexes=indexes)
.
.
.
    def reset_index(
        self,
        dims_or_levels: Union[Hashable, Sequence[Hashable]],
        drop: bool = False,
    ) -> Optional["DataArray"]:
        """Reset the specified index(es) or multi-index level(s).
        Parameters
        ----------
        dims_or_levels : hashable or sequence of hashable
            Name(s) of the dimension(s) and/or multi-index level(s) that will
            be reset.
        drop : bool, optional
            If True, remove the specified indexes and/or multi-index levels
            instead of extracting them as new coordinates (default: False).
        Returns
        -------
        obj : DataArray
            Another dataarray, with this dataarray's data but replaced
            coordinates.
        See Also
        --------
        DataArray.set_index
        """
        coords, _ = split_indexes(
            dims_or_levels, self._coords, set(), self._level_coords, drop=drop
        )
        return self._replace(coords=coords)
```

**Question**:
Do I need to worry that there is any input for `img`, where `img.reset_index` would return `None`?

If I understand everything correctly and this is just a bug, I'm happy to put in a pull request to change the annotation.

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

<!-- Paste the output here xr.show_versions() here -->
INSTALLED VERSIONS
------------------
commit: None
python: 3.9.4 (default, Apr  9 2021, 16:34:09) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 5.4.0-73-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.0
libnetcdf: 4.7.4

xarray: 0.18.2
pandas: 1.2.4
numpy: 1.19.5
scipy: 1.6.1
netCDF4: 1.5.6
pydap: None
h5netcdf: None
h5py: 2.10.0
Nio: None
zarr: None
cftime: 1.5.0
nc_time_axis: None
PseudoNetCDF: None
rasterio: 1.2.4
cfgrib: None
iris: None
bottleneck: None
dask: installed
distributed: None
matplotlib: 3.4.2
cartopy: None
seaborn: None
numbagg: None
pint: None
setuptools: 52.0.0.post20210125
pip: 21.0.1
conda: None
pytest: 6.2.4
IPython: 7.24.1
sphinx: 4.0.2

</details>

--- a/xarray/core/dataarray.py
+++ b/xarray/core/dataarray.py
@@ -1902,7 +1902,7 @@ class DataArray(AbstractArray, DataWithCoords, DataArrayArithmetic):
         indexes: Mapping[Hashable, Union[Hashable, Sequence[Hashable]]] = None,
         append: bool = False,
         **indexes_kwargs: Union[Hashable, Sequence[Hashable]],
-    ) -> Optional["DataArray"]:
+    ) -> "DataArray":
         """Set DataArray (multi-)indexes using one or more existing
         coordinates.
 
@@ -1958,7 +1958,7 @@ class DataArray(AbstractArray, DataWithCoords, DataArrayArithmetic):
         self,
         dims_or_levels: Union[Hashable, Sequence[Hashable]],
         drop: bool = False,
-    ) -> Optional["DataArray"]:
+    ) -> "DataArray":
         """Reset the specified index(es) or multi-index level(s).
 
         Parameters

